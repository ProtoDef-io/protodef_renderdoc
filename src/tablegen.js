import _ from 'lodash';
import assert from 'assert';

export const HORIZONTAL = "horizontal";
export const VERTICAL = "vertical";

export class TableCell {
    
    constructor(slice) {
        if (slice != HORIZONTAL && slice != VERTICAL) { throw "no"; }
        this.slice = slice;
        this.cells = [];

        this.requiredRows = undefined;
        this.requiredCols = undefined;
        this.allocatedRows = undefined;
        this.allocatedCols = undefined;
        this.rowPos = undefined;
        this.colPos = undefined;
    }

    get type() {
        return "cell";
    }

    addCell(slice) {
        let cell = new TableCell(slice);
        this.cells.push(cell);
        return cell;
    }

    addTerminal(inner) {
        let terminal = new TableTerminal(inner);
        this.cells.push(terminal);
    }

    calculateRowCol() {
        let rows = 0;
        let cols = 0;

        if (this.slice == HORIZONTAL) {
            _.each(this.cells, cell => {
                let inner = cell.calculateRowCol();
                rows += inner.rows;
                if (cols < inner.cols) cols = inner.cols;
            });
        } else {
            _.each(this.cells, cell => {
                let inner = cell.calculateRowCol();
                if (rows < inner.rows) rows = inner.rows;
                cols += inner.cols;
            });
        }

        this.requiredRows = rows;
        this.requiredCols = cols;
        return { rows, cols };
    }

    allocateRowCol(rows, cols, rowPos, colPos) {
        this.allocatedRows = rows;
        this.allocatedCols = cols;
        this.rowPos = rowPos;
        this.colPos = colPos;

        assert(this.allocatedRows >= this.requiredRows, "allocated wrong rows");
        assert(this.allocatedCols >= this.requiredCols, "allocated wrong cols");

        if (this.slice == HORIZONTAL) {

            let cellSizesRows = divideEvenly(
                this.allocatedRows - this.requiredRows, 
                this.cells.length
            );

            _.each(_.zip(this.cells, cellSizesRows), ([cell, rowSize]) => {
                let size = cell.requiredRows + rowSize;
                cell.allocateRowCol(size, cols, rowPos, colPos);
                rowPos += size;
            });

        } else {

            let cellSizesCols = divideEvenly(
                this.allocatedCols - this.requiredCols, 
                this.cells.length
            );

            _.each(_.zip(this.cells, cellSizesCols), ([cell, colSize]) => {
                //let size = cell.requiredCols;
                let size = cell.requiredCols + colSize;
                cell.allocateRowCol(rows, size, rowPos, colPos);
                colPos += size;
            });

        }
    }

    toOutput() {
        let { rows, cols } = this.calculateRowCol();
        this.allocateRowCol(rows, cols, 0, 0);

        let out = [];

        for (var cRow = 0; cRow < this.allocatedRows; cRow += 1) {
            let row = [];
            for (var cCol = 0; cCol < this.allocatedCols; cCol += 1) {
                this.rasterize(row, cRow, cCol);
            }
            out.push(row);
        }

        return out;
    }

    rasterize(out, rowPos, colPos) {
        // TODO: Optimize
        _.each(this.cells, cell => cell.rasterize(out, rowPos, colPos));
    }

}

class TableTerminal {

    constructor(inner, parent) {
        this.inner = inner;

        this.requiredRows = 1;
        this.requiredCols = 1;
        this.allocatedRows = undefined;
        this.allocatedCols = undefined;
        this.rowPos = undefined;
        this.colPos = undefined;
    }

    get type() {
        return "terminal";
    }

    calculateRowCol() {
        return { rows: 1, cols: 1 };
    }

    allocateRowCol(rows, cols, rowPos, colPos) {
        this.allocatedRows = rows;
        this.allocatedCols = cols;
        this.rowPos = rowPos;
        this.colPos = colPos;
    }

    rasterize(out, rowPos, colPos) {
        if (this.rowPos == rowPos && this.colPos == colPos) {
            out.push({
                rows: this.allocatedRows,
                cols: this.allocatedCols,
                inner: this.inner,
            });
        }
    }

}

function divideEvenly(number, count) {
    assert(!(number > 0 && count == 0), "could not divideEvenly");
    let elements = _.times(count, _.constant(0));
    while (number > 0) {
        elements = _.map(elements, ec => {
            if (number > 0) {
                number -= 1;
                return ec + 1;
            } else {
                return ec;
            }
        });
    }
    return elements;
}
