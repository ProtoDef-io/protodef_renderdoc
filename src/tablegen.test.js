import { TableCell, HORIZONTAL, VERTICAL } from './tablegen';

test("simple cells", () => {
    let n1 = new TableCell(HORIZONTAL);
    n1.addTerminal("test1");
    let n2 = n1.addCell(VERTICAL);
    n2.addTerminal("test2");
    n2.addTerminal("test3");

    expect(n1.toOutput()).toMatchSnapshot();
});

test("advanced cells 1", () => {
    let n1 = new TableCell(HORIZONTAL);
    n1.addTerminal("test1");
    let n2 = n1.addCell(VERTICAL);
    n2.addTerminal("test2");
    let n3 = n2.addCell(HORIZONTAL);
    n3.addTerminal("test3");
    let n4 = n3.addCell(VERTICAL);
    n4.addTerminal("test4");
    let n5 = n4.addCell(HORIZONTAL);
    n5.addTerminal("test5");
    n5.addTerminal("test6");

    expect(n1.toOutput()).toMatchSnapshot();
});

test("advanced cells 2", () => {
    let n1 = new TableCell(HORIZONTAL);
    n1.addTerminal("test1");
    let n2 = n1.addCell(VERTICAL);
    n2.addTerminal("test2");
    let n3 = n2.addCell(HORIZONTAL);
    n3.addTerminal("test3");
    let n4 = n3.addCell(VERTICAL);
    n4.addTerminal("test4");
    let n5 = n4.addCell(VERTICAL);
    n5.addTerminal("test5");
    n5.addTerminal("test6");

    expect(n1.toOutput()).toMatchSnapshot();
});

test("advanced cells 3", () => {
    let n1 = new TableCell(HORIZONTAL);
    n1.addTerminal("test1");
    let n2 = n1.addCell(VERTICAL);
    n2.addTerminal("test2");
    let n3 = n2.addCell(HORIZONTAL);
    n3.addTerminal("test3");
    let n4 = n3.addCell(VERTICAL);
    n4.addTerminal("test4");
    let n5 = n4.addCell(VERTICAL);
    n5.addTerminal("test5");
    n5.addTerminal("test6");
    n3.addTerminal("test7");

    expect(n1.toOutput()).toMatchSnapshot();
});
