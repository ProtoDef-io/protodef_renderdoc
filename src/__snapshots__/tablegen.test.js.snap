// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`advanced cells 1 1`] = `
Array [
  Array [
    Object {
      "cols": 3,
      "inner": "test1",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test2",
      "rows": 3,
    },
    Object {
      "cols": 2,
      "inner": "test3",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test4",
      "rows": 2,
    },
    Object {
      "cols": 1,
      "inner": "test5",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test6",
      "rows": 1,
    },
  ],
  Array [],
  Array [],
  Array [],
]
`;

exports[`advanced cells 2 1`] = `
Array [
  Array [
    Object {
      "cols": 4,
      "inner": "test1",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test2",
      "rows": 2,
    },
    Object {
      "cols": 3,
      "inner": "test3",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test4",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test5",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test6",
      "rows": 1,
    },
  ],
  Array [],
  Array [],
]
`;

exports[`advanced cells 3 1`] = `
Array [
  Array [
    Object {
      "cols": 4,
      "inner": "test1",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test2",
      "rows": 3,
    },
    Object {
      "cols": 3,
      "inner": "test3",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test4",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test5",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test6",
      "rows": 1,
    },
    Object {
      "cols": 3,
      "inner": "test7",
      "rows": 1,
    },
  ],
  Array [],
  Array [],
  Array [],
]
`;

exports[`simple cells 1`] = `
Array [
  Array [
    Object {
      "cols": 2,
      "inner": "test1",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test2",
      "rows": 1,
    },
    Object {
      "cols": 1,
      "inner": "test3",
      "rows": 1,
    },
  ],
  Array [],
]
`;
